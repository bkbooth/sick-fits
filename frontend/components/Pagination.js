import React from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import Head from 'next/head';
import Link from 'next/link';
import Error from 'components/ErrorMessage';
import PaginationStyles from 'components/styles/PaginationStyles';
import { ITEMS_PER_PAGE } from '../config';

export const PAGINATION_QUERY = gql`
  query PAGINATION {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

const Pagination = ({ page }) => (
  <Query query={PAGINATION_QUERY}>
    {({ data, error, loading }) => {
      if (loading) return <p>Loading...</p>;
      if (error) return <Error error={error} />;
      const { count } = data.itemsConnection.aggregate;
      const pages = Math.ceil(count / ITEMS_PER_PAGE);
      return (
        <PaginationStyles data-test="pagination">
          <Head>
            <title>
              Sick Fits! — Page {page} of {pages}
            </title>
          </Head>
          <Link href={{ pathname: 'items', query: { page: page - 1 } }} prefetch>
            <a className="prev" aria-disabled={page <= 1}>
              ← Prev
            </a>
          </Link>
          <p>
            Page {page} of <span data-test="total-pages">{pages}</span>
          </p>
          <p>{count} items total</p>
          <Link href={{ pathname: 'items', query: { page: page + 1 } }} prefetch>
            <a className="next" aria-disabled={page >= pages}>
              Next →
            </a>
          </Link>
        </PaginationStyles>
      );
    }}
  </Query>
);

export default Pagination;
