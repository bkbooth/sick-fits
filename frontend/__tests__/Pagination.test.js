import { MockedProvider } from 'react-apollo/test-utils';
import { mount } from 'enzyme';
import wait from 'waait';
import Pagination, { PAGINATION_QUERY } from 'components/Pagination';

function makeMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          itemsConnection: {
            __typename: 'Aggregate',
            aggregate: { __typename: 'Count', count: length },
          },
        },
      },
    },
  ];
}

describe('<Nav />', () => {
  it('displays a loading message', () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(1)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    expect(wrapper.text()).toContain('Loading...');
  });

  it('renders pagination for 18 items', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('[data-test="total-pages"]').text()).toBe('5');
    expect(wrapper.find('div[data-test="pagination"]')).toMatchSnapshot();
  });

  it('disables prev button on the first page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('a.prev').prop('aria-disabled')).toBe(true);
    expect(wrapper.find('a.next').prop('aria-disabled')).toBe(false);
  });

  it('disables next button on the last page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={5} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('a.prev').prop('aria-disabled')).toBe(false);
    expect(wrapper.find('a.next').prop('aria-disabled')).toBe(true);
  });

  it('enables all buttons on a middle page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={3} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('a.prev').prop('aria-disabled')).toBe(false);
    expect(wrapper.find('a.next').prop('aria-disabled')).toBe(false);
  });
});
