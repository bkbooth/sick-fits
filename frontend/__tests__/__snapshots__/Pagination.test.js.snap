// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> renders pagination for 18 items 1`] = `
<div
  className="PaginationStyles-aduuar-0 ewJsCc"
  data-test="pagination"
>
  <Head>
    <Component
      handleStateChange={null}
      inAmpMode={false}
      reduceComponentsToState={[Function]}
    />
  </Head>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 0,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={true}
      className="prev"
      href="items?page=0"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      ← Prev
    </a>
  </Link>
  <p>
    Page 
    1
     of 
    <span
      data-test="total-pages"
    >
      5
    </span>
  </p>
  <p>
    18
     items total
  </p>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 2,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={false}
      className="next"
      href="items?page=2"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Next →
    </a>
  </Link>
</div>
`;
